File: system_tests/tests/visualizer_system_test.py
"""
System tests for the visualizer service – Python client (native/pybind11).

Frame combinations are loaded from system_tests/visualizer_frames.yaml: every source frame is tested against every
target frame.
"""

import itertools
import math
import os
import time

import numpy as np
import pytest
import yaml
from planning_service_client.native.types import Conf, SystemConf, VisualizerStatus
from planning_service_client.native.visualizer import VisualizerClient

VIZ_ADDR = os.getenv("VIZ_ADDR", "visualizer-dev:5550")

# Locate visualizer_frames.yaml whether running under Bazel or directly.
_TEST_SRCDIR = os.getenv("TEST_SRCDIR")
if _TEST_SRCDIR:
    _FRAMES_FILE = os.path.join(_TEST_SRCDIR, "_main", "system_tests", "visualizer_frames.yaml")
else:
    _FRAMES_FILE = os.path.join(os.path.dirname(__file__), "..", "visualizer_frames.yaml")

with open(_FRAMES_FILE) as _f:
    _cfg = yaml.safe_load(_f)
    _FRAME_PAIRS = list(itertools.product(_cfg["source_frames"], _cfg["target_frames"]))


# ---------------------------------------------------------------------------
# Helpers
# ---------------------------------------------------------------------------


def _wait_for_active(client: VisualizerClient, *, timeout_s: float = 60.0, poll_s: float = 1.0):
    """
    Poll get_visualizer_status until the session is ACTIVE.
    """
    deadline = time.monotonic() + timeout_s
    while time.monotonic() < deadline:
        s = client.get_visualizer_status()
        if s.status == VisualizerStatus.Status.ACTIVE:
            return s
        time.sleep(poll_s)
    raise TimeoutError(f"Visualizer did not reach ACTIVE status within {timeout_s:.0f} s")


# ---------------------------------------------------------------------------
# Module-scoped client – one connection for the whole file
# ---------------------------------------------------------------------------


@pytest.fixture(scope="module")
def viz_client() -> VisualizerClient:
    client = VisualizerClient(VIZ_ADDR, "py_viz_system_test")
    # Wait 30 seconds max
    assert client.connect(num_attempts=30, attempt_interval_ms=1000), "Could not connect to visualizer service"
    _wait_for_active(client)
    return client


# ---------------------------------------------------------------------------
# Tests
# ---------------------------------------------------------------------------


def test_is_active(viz_client):
    """
    The service should have loaded the default lifty context on startup.
    """
    assert viz_client.get_visualizer_status().status == VisualizerStatus.Status.ACTIVE


@pytest.mark.parametrize(
    "frame_a,frame_b",
    _FRAME_PAIRS,
    ids=[f"{a}__{b}".replace("::", "_") for a, b in _FRAME_PAIRS],
)
def test_calc_pose(viz_client, frame_a, frame_b):
    """
    calc_pose should return a valid FrameRelativePose for every combination.
    """
    pose = viz_client.calc_pose(frame_a, frame_b)
    assert pose.frame_A == frame_a
    assert pose.frame_B == frame_b
    assert pose.translation is not None
    assert pose.quaternion_wxyz is not None


# ---------------------------------------------------------------------------
# Joint-position override for calc_pose
# ---------------------------------------------------------------------------


class TestCalcPoseJointOverride:
    """
    Tests for the system_conf_override parameter of calc_pose.

    Lifty joint order within each Conf vector: [rev, piston]
      rev    ∈ [-1.57, 1.57]  (revolute)
      piston ∈ [0.00,  0.30]  (prismatic)
    """

    _CONF_A = [1.0, 0.25]  # a: rev=1.0, piston=0.25
    _CONF_B = [-1.0, 0.05]  # b: rev=-1.0, piston=0.05

    @staticmethod
    def _make_override() -> SystemConf:
        sc = SystemConf()
        sc.set("a", Conf(np.array(TestCalcPoseJointOverride._CONF_A)))
        sc.set("b", Conf(np.array(TestCalcPoseJointOverride._CONF_B)))
        return sc

    @pytest.mark.parametrize(
        "frame_a,frame_b",
        _FRAME_PAIRS,
        ids=[f"{a}__{b}".replace("::", "_") for a, b in _FRAME_PAIRS],
    )
    def test_returns_finite_well_formed_pose(self, viz_client, frame_a, frame_b):
        """
        calc_pose with an override must return a finite, well-formed pose for every frame combination.
        """
        pose = viz_client.calc_pose(frame_a, frame_b, self._make_override())
        assert pose.frame_A == frame_a
        assert pose.frame_B == frame_b
        assert all(math.isfinite(v) for v in pose.translation)
        assert all(math.isfinite(v) for v in pose.quaternion_wxyz)
        norm_sq = sum(v**2 for v in pose.quaternion_wxyz)
        assert abs(norm_sq - 1.0) < 1e-6

    def test_pose_differs_from_default(self, viz_client):
        """
        The override must actually change the returned pose.
        """
        frame_a = "world"
        frame_b = "a::stamp"
        default_pose = viz_client.calc_pose(frame_a, frame_b)
        overridden_pose = viz_client.calc_pose(frame_a, frame_b, self._make_override())
        assert not np.allclose(
            np.array(default_pose.translation),
            np.array(overridden_pose.translation),
            atol=1e-3,
        ), "Pose did not change when a joint-position override was applied"


# ---------------------------------------------------------------------------
# toggle_frame — path resolution
# ---------------------------------------------------------------------------

# A frame that is guaranteed to exist in the loaded lifty model.
_KNOWN_FRAME = "a::stamp"
# Its canonical Meshcat path (model instance "a", frame "stamp").
_KNOWN_FRAME_ABS_PATH = "/drake/frames/a/stamp"


@pytest.mark.parametrize("visible", [True, False], ids=["on", "off"])
def test_toggle_frame_bare_name(viz_client, visible):
    """
    Case 3: bare name (no '/') is resolved via the loaded model.
    """
    viz_client.toggle_frame(_KNOWN_FRAME, visible)


@pytest.mark.parametrize("visible", [True, False], ids=["on", "off"])
def test_toggle_frame_absolute_path(viz_client, visible):
    """
    Case 1: absolute path under /drake/frames/ is used verbatim.
    """
    viz_client.toggle_frame(_KNOWN_FRAME_ABS_PATH, visible)


@pytest.mark.parametrize("visible", [True, False], ids=["on", "off"])
def test_toggle_frame_relative_with_frames_prefix(viz_client, visible):
    """
    Case 2a: relative path starting with "frames/" – "/drake/" is prepended.
    Equivalent to _KNOWN_FRAME_ABS_PATH.
    """
    viz_client.toggle_frame("frames/a/stamp", visible)


@pytest.mark.parametrize("visible", [True, False], ids=["on", "off"])
def test_toggle_frame_relative_without_frames_prefix(viz_client, visible):
    """
    Case 2b: relative path without "frames/" prefix – "/drake/frames/" is
    prepended.  "a/stamp" → /drake/frames/a/stamp/.
    """
    viz_client.toggle_frame("a/stamp", visible)


def test_toggle_frame_roundtrip(viz_client):
    """
    Toggling off then on should not raise for any supported input form.
    """
    for frame in (_KNOWN_FRAME, _KNOWN_FRAME_ABS_PATH, "frames/a/stamp", "a/stamp"):
        viz_client.toggle_frame(frame, False)
        viz_client.toggle_frame(frame, True)


def test_toggle_frame_empty_name_raises(viz_client):
    """
    An empty frame string must be rejected by the server.
    """
    with pytest.raises(RuntimeError):
        viz_client.toggle_frame("", True)


def test_toggle_frame_wrong_subtree_raises(viz_client):
    """
    An absolute path not under /drake/frames/ must be rejected.
    """
    with pytest.raises(RuntimeError):
        viz_client.toggle_frame("/drake/objects/some_object", True)


def test_toggle_frame_unknown_bare_name_raises(viz_client):
    """
    A bare name that does not exist in the loaded model must be rejected.
    """
    with pytest.raises(RuntimeError):
        viz_client.toggle_frame("this_frame_does_not_exist_xyz", True)

File: system_tests/cpp/visualizer_system_test.cc
/*
 * System test for the visualizer service – C++ client (GoogleTest).
 *
 * Frame combinations are loaded from system_tests/visualizer_frames.yaml.
 * Every source frame is tested against every target frame.
 *
 * Configuration via environment variables:
 *   VIZ_ADDR    host:port of the visualizer service (default:
 * visualizer-dev:5550) TEST_SRCDIR set by Bazel; used to locate the YAML data
 * file
 */

#include <drake/common/name_value.h>
#include <drake/common/yaml/yaml_io.h>
#include <gtest/gtest.h>

#include <chrono>
#include <cstdlib>
#include <memory>
#include <string>
#include <thread>
#include <tuple>
#include <vector>

#include "planning_service_client/api/visualizer_client.h"
#include "planning_service_client/conf.h"
#include "planning_service_client/frame_relative_pose.h"
#include "planning_service_client/visualizer_status.h"

namespace psc = planning_service_client;

// ---------------------------------------------------------------------------
// YAML config
// ---------------------------------------------------------------------------

struct VisualizerTestConfig {
  std::vector<std::string> source_frames;
  std::vector<std::string> target_frames;
  template <typename Archive>
  void Serialize(Archive* a) {
    a->Visit(DRAKE_NVP(source_frames));
    a->Visit(DRAKE_NVP(target_frames));
  }
};

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

static std::string EnvOr(const char* name, const char* fallback) {
  const char* v = std::getenv(name);
  return v ? std::string(v) : std::string(fallback);
}

static std::string FramesFilePath() {
  const char* srcdir = std::getenv("TEST_SRCDIR");
  if (srcdir) {
    return std::string(srcdir) + "/_main/system_tests/visualizer_frames.yaml";
  }
  return "system_tests/visualizer_frames.yaml";
}

// Returns the cartesian product of source_frames × target_frames.
static std::vector<std::tuple<std::string, std::string>> LoadFramePairs() {
  auto cfg = drake::yaml::LoadYamlFile<VisualizerTestConfig>(FramesFilePath());
  std::vector<std::tuple<std::string, std::string>> pairs;
  for (const auto& src : cfg.source_frames) {
    for (const auto& tgt : cfg.target_frames) {
      pairs.emplace_back(src, tgt);
    }
  }
  return pairs;
}

// ---------------------------------------------------------------------------
// Helper: build a SystemConf override for the two lifty robot instances.
//
// Lifty joint order within each Conf vector: [rev, piston]
//   rev    ∈ [-1.57, 1.57]  (revolute)
//   piston ∈ [0.00,  0.30]  (prismatic)
// ---------------------------------------------------------------------------
static psc::SystemConf MakeOverrideConf() {
  psc::SystemConf sc;
  Eigen::VectorXd qa(2);
  qa << 1.0, 0.25;  // a: rev=1.0, piston=0.25
  sc["a"] = psc::Conf(qa);
  Eigen::VectorXd qb(2);
  qb << -1.0, 0.05;  // b: rev=-1.0, piston=0.05
  sc["b"] = psc::Conf(qb);
  return sc;
}

static psc::VisualizerStatus WaitForActive(
    const psc::client::VisualizerClient& client, int timeout_s = 60,
    int poll_ms = 1000) {
  const auto deadline =
      std::chrono::steady_clock::now() + std::chrono::seconds(timeout_s);
  while (std::chrono::steady_clock::now() < deadline) {
    auto s = client.GetVisualizerStatus();
    if (s.status() == psc::VisualizerStatus::Status::kActive) return s;
    std::this_thread::sleep_for(std::chrono::milliseconds(poll_ms));
  }
  throw std::runtime_error("Timed out waiting for ACTIVE status");
}

// ---------------------------------------------------------------------------
// Shared client – created once for the whole process
// ---------------------------------------------------------------------------

static std::unique_ptr<psc::client::VisualizerClient> g_client;

// ---------------------------------------------------------------------------
// Lifecycle test (runs first – alphabetically before "Lifty/...")
// ---------------------------------------------------------------------------

class VisualizerLifecycleTest : public ::testing::Test {
 protected:
  static void SetUpTestSuite() {
    g_client = std::make_unique<psc::client::VisualizerClient>(
        EnvOr("VIZ_ADDR", "visualizer-dev:5550"), "cpp_viz_system_test");
    ASSERT_TRUE(
        g_client->Connect(/*num_attempts=*/60, /*attempt_interval_ms=*/1000))
        << "Could not connect to visualizer service";
  }
};

TEST_F(VisualizerLifecycleTest, DefaultContextIsActive) {
  ASSERT_NO_THROW(WaitForActive(*g_client, 60));
}

// ---------------------------------------------------------------------------
// Parameterised CalcPose tests – one per (source, target) combination
// ---------------------------------------------------------------------------

class VisualizerCalcPoseTest
    : public ::testing::TestWithParam<std::tuple<std::string, std::string>> {};

TEST_P(VisualizerCalcPoseTest, CalcPose) {
  const auto& [frame_a, frame_b] = GetParam();
  psc::FrameRelativePose pose;
  ASSERT_NO_THROW(pose = g_client->CalcPose(frame_a, frame_b));
  EXPECT_EQ(pose.frame_A(), frame_a);
  EXPECT_EQ(pose.frame_B(), frame_b);
}

INSTANTIATE_TEST_SUITE_P(
    Lifty, VisualizerCalcPoseTest, ::testing::ValuesIn(LoadFramePairs()),
    [](const ::testing::TestParamInfo<std::tuple<std::string, std::string>>&
           info) {
      // Replace non-alphanumeric chars to get a valid name
      auto sanitise = [](std::string s) {
        for (char& c : s)
          if (!std::isalnum(c)) c = '_';
        return s;
      };
      return sanitise(std::get<0>(info.param)) + "__"
             + sanitise(std::get<1>(info.param));
    });

// ---------------------------------------------------------------------------
// CalcPose joint-override tests
//
// VisualizerCalcPoseOverrideBase  – shared fixture; holds MakeOverrideConf()
//   ├─ TEST_F  PoseDiffersFromDefault   (single non-parameterised test)
//   └─ VisualizerCalcPoseOverrideTest   (parameterised over all frame pairs)
//        └─ TEST_P  CalcPoseWithOverride
// ---------------------------------------------------------------------------

class VisualizerCalcPoseOverrideBase : public ::testing::Test {
 protected:
  static psc::SystemConf MakeOverrideConf() {
    return ::MakeOverrideConf();
  }
};

// Single test: the override must actually change the returned pose.
TEST_F(VisualizerCalcPoseOverrideBase, PoseDiffersFromDefault) {
  const psc::FrameRelativePose default_pose =
      g_client->CalcPose("a::stamp", "world");
  const psc::FrameRelativePose overridden_pose =
      g_client->CalcPose("a::stamp", "world", MakeOverrideConf());
  EXPECT_FALSE(default_pose.X_AB_translation().isApprox(
      overridden_pose.X_AB_translation(), 1e-3))
      << "Pose did not change when a joint-position override was applied";
}

// Parameterised tests: well-formed pose for every (source, target) pair.
class VisualizerCalcPoseOverrideTest
    : public VisualizerCalcPoseOverrideBase,
      public ::testing::WithParamInterface<
          std::tuple<std::string, std::string>> {};

TEST_P(VisualizerCalcPoseOverrideTest, CalcPoseWithOverride) {
  const auto& [frame_a, frame_b] = GetParam();
  psc::FrameRelativePose pose;
  ASSERT_NO_THROW(pose =
                      g_client->CalcPose(frame_a, frame_b, MakeOverrideConf()));
  EXPECT_EQ(pose.frame_A(), frame_a);
  EXPECT_EQ(pose.frame_B(), frame_b);
  EXPECT_TRUE(pose.X_AB_translation().allFinite());
  EXPECT_TRUE(pose.X_AB_quaternion().coeffs().allFinite());
  EXPECT_NEAR(pose.X_AB_quaternion().norm(), 1.0, 1e-6);
}

INSTANTIATE_TEST_SUITE_P(Lifty, VisualizerCalcPoseOverrideTest,
                         ::testing::ValuesIn(LoadFramePairs()),
                         [](const ::testing::TestParamInfo<
                             std::tuple<std::string, std::string>>& info) {
                           auto sanitise = [](std::string s) {
                             for (char& c : s)
                               if (!std::isalnum(c)) c = '_';
                             return s;
                           };
                           return sanitise(std::get<0>(info.param)) + "__"
                                  + sanitise(std::get<1>(info.param));
                         });

// ---------------------------------------------------------------------------
// ToggleFrame – path resolution integration tests
// ---------------------------------------------------------------------------

// A frame guaranteed to exist in the loaded lifty model.
static constexpr std::string_view kKnownFrame {"a::stamp"};
// Its canonical Meshcat path.
static constexpr std::string_view kKnownFrameAbsPath {"/drake/frames/a/stamp"};

class VisualizerToggleFrameTest : public ::testing::Test {};

// Case 3 — bare name resolved via the loaded model.
TEST_F(VisualizerToggleFrameTest, BareNameHide) {
  EXPECT_NO_THROW(g_client->ToggleFrame(kKnownFrame, false));
}
TEST_F(VisualizerToggleFrameTest, BareNameShow) {
  EXPECT_NO_THROW(g_client->ToggleFrame(kKnownFrame, true));
}

// Case 1 — absolute path under /drake/frames/ used verbatim.
TEST_F(VisualizerToggleFrameTest, AbsolutePathHide) {
  EXPECT_NO_THROW(g_client->ToggleFrame(kKnownFrameAbsPath, false));
}
TEST_F(VisualizerToggleFrameTest, AbsolutePathShow) {
  EXPECT_NO_THROW(g_client->ToggleFrame(kKnownFrameAbsPath, true));
}

// Case 2a — relative path with "frames/" prefix: "/drake/" is prepended.
TEST_F(VisualizerToggleFrameTest, RelativeFramesPrefixHide) {
  EXPECT_NO_THROW(g_client->ToggleFrame("frames/a/stamp", false));
}
TEST_F(VisualizerToggleFrameTest, RelativeFramesPrefixShow) {
  EXPECT_NO_THROW(g_client->ToggleFrame("frames/a/stamp", true));
}

// Case 2b — relative path without "frames/" prefix: "/drake/frames/" prepended.
TEST_F(VisualizerToggleFrameTest, RelativeNoFramesPrefixHide) {
  EXPECT_NO_THROW(g_client->ToggleFrame("a/stamp", false));
}
TEST_F(VisualizerToggleFrameTest, RelativeNoFramesPrefixShow) {
  EXPECT_NO_THROW(g_client->ToggleFrame("a/stamp", true));
}

// Roundtrip — all four input forms: hide then show.
TEST_F(VisualizerToggleFrameTest, Roundtrip) {
  for (std::string_view frame :
       {kKnownFrame, kKnownFrameAbsPath, std::string_view {"frames/a/stamp"},
        std::string_view {"a/stamp"}}) {
    EXPECT_NO_THROW(g_client->ToggleFrame(frame, false)) << "frame=" << frame;
    EXPECT_NO_THROW(g_client->ToggleFrame(frame, true)) << "frame=" << frame;
  }
}

// Error cases — server must reject these.
TEST_F(VisualizerToggleFrameTest, EmptyNameThrows) {
  EXPECT_THROW(g_client->ToggleFrame("", true), std::runtime_error);
}

TEST_F(VisualizerToggleFrameTest, WrongSubtreeThrows) {
  EXPECT_THROW(g_client->ToggleFrame("/drake/objects/some_object", true),
               std::runtime_error);
}

TEST_F(VisualizerToggleFrameTest, UnknownBareNameThrows) {
  EXPECT_THROW(g_client->ToggleFrame("this_frame_does_not_exist_xyz", true),
               std::runtime_error);
}
